### ISSUE TITLE
Groovy compiler error: exception in phase 'semantic analysis'

### ISSUE THREAD

**Issue by ctgnz** [groovy/groovy-eclipse#868]
I got a compilation error saying "Groovy compiler error: exception in phase 'semantic analysis'"

On inspecting the logs (see below), I discovered this was at a piece of code invoking the Spring REST template to call a REST method that returns a List of Strings. The REST template needs a parameterized type reference to be able to cast to List<String>, so I was doing that directly in the call to template.exchange using an anonymous inner class:

`template.exchange(uri, HttpMethod.GET, null,  new ParameterizedTypeReference<List<String>>() {})`

If I extract the type reference into a constant, the compile error goes away:

```
static final def TYPE_STRING_LIST = new ParameterizedTypeReference<List<String>>() {}
template.exchange(uri, HttpMethod.GET, null,  TYPE_STRING_LIST)
```

I am running GRECLIPSE 3.3.0 in Eclipse 2019-03, using Groovy 2.4.16 compiler.

Here's the error from the log:

```
!ENTRY org.eclipse.jdt.groovy.core 4 0 2019-04-09 12:00:08.982
!MESSAGE Groovy compiler error
!STACK 0
BUG! exception in phase 'semantic analysis' in source unit 'MyClass.groovy' commencingResolution failed: declaration found, but unexpectedly found no scope for MyClass$2
        at org.codehaus.groovy.control.CompilationUnit.applyToSourceUnits(CompilationUnit.java:989)
        at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:632)
        at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:581)
        at org.codehaus.jdt.groovy.internal.compiler.ast.GroovyCompilationUnitDeclaration.processToPhase(GroovyCompilationUnitDeclaration.ja
va:222)
        at org.codehaus.jdt.groovy.internal.compiler.ast.GroovyCompilationUnitDeclaration.resolve(GroovyCompilationUnitDeclaration.java:606)
        at org.eclipse.jdt.internal.compiler.Compiler.resolve(Compiler.java:1071)
        at org.eclipse.jdt.internal.compiler.Compiler.resolve(Compiler.java:1116)
        at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:291)
        at org.eclipse.jdt.internal.core.CompilationUnitProblemFinder.process(CompilationUnitProblemFinder.java:358)
        at org.codehaus.jdt.groovy.model.GroovyReconcileWorkingCopyOperation.makeConsistent(GroovyReconcileWorkingCopyOperation.java:82)
        at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.executeOperation(ReconcileWorkingCopyOperation.java:94)
        at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:736)
        at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:802)
        at org.codehaus.jdt.groovy.model.GroovyCompilationUnit.reconcile(GroovyCompilationUnit.java:427)
        at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:131)
        at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.access$0(JavaReconcilingStrategy.java:113)
        at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy$1.run(JavaReconcilingStrategy.java:93)
        at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
        at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:90)
        at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:157)
        at org.eclipse.jdt.internal.ui.text.CompositeReconcilingStrategy.reconcile(CompositeReconcilingStrategy.java:92)
        at org.eclipse.jdt.internal.ui.text.JavaCompositeReconcilingStrategy.reconcile(JavaCompositeReconcilingStrategy.java:107)
        at org.eclipse.jface.text.reconciler.MonoReconciler.process(MonoReconciler.java:76)
        at org.eclipse.jface.text.reconciler.AbstractReconciler$BackgroundThread.run(AbstractReconciler.java:210)
Caused by: org.codehaus.jdt.groovy.internal.compiler.ast.GroovyEclipseBug: commencingResolution failed: declaration found, but unexpectedly 
found no scope for MyClass$2
        at org.codehaus.jdt.groovy.internal.compiler.ast.JDTResolver.commencingResolution(JDTResolver.java:229)
        at org.codehaus.groovy.control.ResolveVisitor.visitClass(ResolveVisitor.java:1441)
        at org.codehaus.groovy.control.ResolveVisitor.startResolving(ResolveVisitor.java:228)
        at org.codehaus.jdt.groovy.internal.compiler.ast.JDTResolver.startResolving(JDTResolver.java:191)
        at org.codehaus.groovy.control.CompilationUnit$1.call(CompilationUnit.java:692)
        at org.codehaus.groovy.control.CompilationUnit.applyToSourceUnits(CompilationUnit.java:980)
        ... 23 more

```

**Comment by eric-milles** [comment]
Would you mind sharing more of the class that has the error?  I am not seeing the same with a small test case:
```groovy
import org.springframework.core.ParameterizedTypeReference
import org.springframework.http.HttpMethod
import org.springframework.web.client.RestTemplate

def template = new RestTemplate(); URI uri = null
def response = template.exchange(uri, HttpMethod.GET, null, new ParameterizedTypeReference<List<String>>() {})
```

Also, have you tried the latest snapshot build?  There are a few fixes since 3.3.0 available.

**Comment by eric-milles** [comment]
One reason that this error can occur is if the containing class has errors.  You can open the Console view and from there open the Groovy Event Console from the last button in the toolbar of the view.  Then you can close and reopen your source file to see it reconciled (parsed and processed for editing).  If you have auto build on, you can make one change (like adding a space) and save to see incremental compilation.
